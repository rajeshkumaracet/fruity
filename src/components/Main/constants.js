export const INCOME_STATEMENT_URL =
  "https://www.alphavantage.co/query?function=INCOME_STATEMENT&symbol=IBM&apikey=demo";
export const BALANCE_SHEET_URL =
  "https://www.alphavantage.co/query?function=BALANCE_SHEET&symbol=IBM&apikey=demo";
