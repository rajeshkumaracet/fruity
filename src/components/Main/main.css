.main {
  margin: 20px;
}

.card {
  height: 100vh;
  text-align: center;
  border-radius: 10px;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
  transition: box-shadow 0.3s ease;
}

.card:hover {
  box-shadow: 0 0 20px rgba(0, 0, 0, 0.2);
}

.error {
  height: 70%;
  width: 100%;
  display: flex;
  position: fixed;
  align-items: center;
  justify-content: center;
}

.danger-icon {
  font-size: 100px;
  display: block;
  align-self: center;
}

.error-label {
  font-size: 30px;
  margin-top: 30px;
  text-transform: uppercase;
}
