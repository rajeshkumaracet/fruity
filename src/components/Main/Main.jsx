import { useEffect, useState } from "react";
import axios from "axios";
import { Col, Row, Card } from "antd";
import { ExclamationCircleTwoTone } from "@ant-design/icons";

import PlotChart from "../Chart/Chart";

import "./Main.css";

import { BALANCE_SHEET_URL, INCOME_STATEMENT_URL } from "./constants";

const Main = () => {
  const [fetching, setFetching] = useState(false);
  const [error, setError] = useState(false);
  const [data, setData] = useState(null);

  const fetchData = async () => {
    try {
      setFetching(true);
      const getBalanceSheet = await axios.get(`${BALANCE_SHEET_URL}`);
      const getIncomeStatement = await axios.get(`${INCOME_STATEMENT_URL}`);
      setData({
        balanceSheetData: getBalanceSheet?.data?.quarterlyReports,
        incomeStatementData: getIncomeStatement?.data?.quarterlyReports,
        symbol: getBalanceSheet?.data?.symbol,
      });
      setFetching(false);
    } catch (error) {
      setFetching(false);
      setError(true);
    }
  };

  useEffect(() => {
    fetchData();
  }, []);

  const Error = () => (
    <div className="error" data-testid="error">
      <div>
        <ExclamationCircleTwoTone twoToneColor="red" className="danger-icon" />
        <p className="error-label">Something wents wrong!</p>
      </div>
    </div>
  );

  const Graph = () => (
    <div data-testid="graph">
      <Row>
        <Col className="card1" span={24}>
          <Card className="card" loading={fetching}>
            {data && <PlotChart data={data} />}
          </Card>
        </Col>
      </Row>
    </div>
  );

  return (
    <div className="main" data-testid="main">
      {error ? <Error /> : <Graph />}
    </div>
  );
};

export default Main;
