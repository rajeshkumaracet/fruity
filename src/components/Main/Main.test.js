import React from "react";
import { render, waitFor, screen } from "@testing-library/react";
import axios from "axios";
import Main from "./Main";

jest.mock("axios"); // Mock axios module

describe("<Main/>", () => {
  it("should render correctly", () => {
    const { container } = render(<Main />);
    expect(container).toMatchSnapshot();
  });

  it("renders graph when data is fetched successfully", async () => {
    const balanceSheetData = [
      { id: 1, date: "2022-01-01", value: 100 },
      { id: 2, date: "2022-04-01", value: 200 },
    ];
    const incomeStatementData = [
      { id: 1, date: "2022-01-01", value: 50 },
      { id: 2, date: "2022-04-01", value: 150 },
    ];
    const symbol = "AAPL";

    axios.get.mockImplementationOnce((url) => {
      if (url === "BALANCE_SHEET_URL") {
        return Promise.resolve({
          data: { quarterlyReports: balanceSheetData, symbol },
        });
      } else if (url === "INCOME_STATEMENT_URL") {
        return Promise.resolve({
          data: { quarterlyReports: incomeStatementData },
        });
      }
    });

    render(<Main />);

    await waitFor(() => {
      expect(screen.getByTestId("main")).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.getByTestId("graph")).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.queryByTestId("error")).not.toBeInTheDocument();
    });
  });

  it("displays error message when data fetching fails", async () => {
    axios.get.mockRejectedValueOnce(new Error("Failed to fetch data"));

    render(<Main />);

    await waitFor(() => {
      expect(screen.getByTestId("main")).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.getByTestId("error")).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.queryByTestId("graph")).not.toBeInTheDocument();
    });
  });
});
