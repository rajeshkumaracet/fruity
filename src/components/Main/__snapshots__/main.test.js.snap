// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main/> should render correctly 1`] = `
<div>
  <div
    class="main"
    data-testid="main"
  >
    <div
      data-testid="graph"
    >
      <div
        class="ant-row css-dev-only-do-not-override-1kuana8"
      >
        <div
          class="ant-col ant-col-24 card1 css-dev-only-do-not-override-1kuana8"
        >
          <div
            class="ant-card ant-card-loading ant-card-bordered card css-dev-only-do-not-override-1kuana8"
          >
            <div
              class="ant-card-body"
            >
              <div
                class="ant-skeleton ant-skeleton-active css-dev-only-do-not-override-1kuana8"
              >
                <div
                  class="ant-skeleton-content"
                >
                  <ul
                    class="ant-skeleton-paragraph"
                  >
                    <li />
                    <li />
                    <li />
                    <li
                      style="width: 61%;"
                    />
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
